### PR TITLE
fix: Consolidate pod condition check

### PR DESCRIPTION
**Pull request by Elsie4ever**
1. Consolidate pod condition check
    Add only one evaluation for one pod instead of multiple eval
    summary level: show one line summary of pod health, the pod health can be Healthy/Unhealthy/Indeterminate
    image
    image

    detail level: show one line summary of pod condition + failed check
    image
    image

    verbose level: show everything
    image
    image

example of indeterminate state:
image

  **TODO: will consolidate more as a table view in next release**
image
image

2. Also add fallback warning when there is a pod condition not in known pod condition enum, when all the known pod conditions are in good state
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to Azure IoT Operations tooling!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [ ] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [ ] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [ ] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [ ] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT Operations CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

**Comment by Elsie4ever** [comment]
Unit test + IT test run
image

File: azext_edge/edge/providers/check/base/pod.py [hunk 1]
```diff
@@ -81,11 +82,7 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
 
         pod_conditions = [
             f"{target_service_pod}.status.phase",
-            f"{target_service_pod}.status.conditions.ready",
-            f"{target_service_pod}.status.conditions.initialized",
-            f"{target_service_pod}.status.conditions.containersready",
-            f"{target_service_pod}.status.conditions.podscheduled",
-            f"{target_service_pod}.status.conditions.podreadytostartcontainers",
+            f"{target_service_pod}.status.conditions",
```
Comment (vilit1) [inline azext_edge/edge/providers/check/base/pod.py:85]: personally would like to have an even simpler one line for the entire pod state (is it good? is it bad?) in summary view

But if we want to show the phase vs condition - this is ok

File: azext_edge/edge/providers/check/base/pod.py [hunk 2]
```diff
@@ -121,40 +114,91 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
...
+
+            for condition in pod_conditions:
+                type = condition["type"]
+
+                try:
+                    condition_type = POD_CONDITION_TEXT_MAP[type]
+                    condition_status = condition.get("status") == "True"
+                    conditions_readiness = conditions_readiness and condition_status
+                    pod_condition_deco, status = _decorate_pod_condition(condition=condition_status)
+                except KeyError:
```
Comment (vilit1) [inline azext_edge/edge/providers/check/base/pod.py:129]: Key error catching could be avoided with a get (and then an if none)

Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py:129]: but we want a warning when the pod condition is non regular enum?

File: azext_edge/edge/providers/check/base/pod.py [hunk 3]
```diff
@@ -121,40 +114,91 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
...
+                    condition_type = type
+                    condition_status = condition.get("status")
+
+                formatted_reason = ""
+                condition_reason = condition.get("reason", "")
+
+                if condition_reason:
+                    # remove the [ and ] to prevent console not printing the text
+                    condition_reason = condition_reason.replace("[", "\\[")
```
Comment (vilit1) [inline azext_edge/edge/providers/check/base/pod.py:138]: did you get the ] here too? or does is the rich linter smart enough to skip ]'s if it cannot find an opening [?

File: azext_edge/edge/providers/check/base/pod.py [hunk 4]
```diff
@@ -121,40 +114,91 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
...
-                            ),
+                            display=Padding(reason, (0, 0, 0, padding + 8)),
                         )
+
+            if not conditions_readiness:
+                pod_eval_status = CheckTaskStatus.error.value
+            else:
+                # add warning if there are unknown conditions when known conditions are all in good state
+                for condition, reason in unknown_conditions_display_list:
+                    condition_text: str = f"[yellow]Irrgular Condition {condition} found.[/yellow]"
```
Comment (vilit1) [inline azext_edge/edge/providers/check/base/pod.py:182]: irregular

File: azext_edge/edge/providers/check/base/pod.py [hunk 5]
```diff
@@ -121,40 +114,91 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
...
+
+                formatted_reason = ""
+                condition_reason = condition.get("reason", "")
+
+                if condition_reason:
+                    # remove the [ and ] to prevent console not printing the text
+                    condition_reason = condition_reason.replace("[", "\\[")
+                    formatted_reason = f"[red]Reason: {condition_reason}[/red]"
+
+                known_condition_values = [value.replace(" ", "").lower() for value in POD_CONDITION_TEXT_MAP.values()]
```
Comment (vilit1) [inline azext_edge/edge/providers/check/base/pod.py:141]: can we not trust the pod condition to have spaces correctly?

File: azext_edge/edge/providers/check/base/pod.py (hunk 2 above)
Comment (vilit1) [reply azext_edge/edge/providers/check/base/pod.py:129]: yes - what I meant here is don't do try except if you can do if's easily

File: azext_edge/edge/providers/check/base/pod.py (hunk 5 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py:127]: no we can't😣

File: azext_edge/edge/providers/check/base/pod.py (hunk 1 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py:85]: updated~

File: azext_edge/edge/providers/check/base/pod.py (hunk 2 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py]: updated

File: azext_edge/edge/providers/check/base/pod.py (hunk 5 above)
Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py:127]: quote your meme
image

File: azext_edge/edge/providers/check/base/pod.py [hunk 6]
```diff
@@ -96,65 +93,159 @@ def _decorate_pod_condition(condition: bool) -> Tuple[str, str]:
...
+        pod_eval_value["status.phase"] = pod_phase
+
+        conditions_readiness = True
+        conditions_display_list: List[Tuple[str, str]] = []
+        unknown_conditions_display_list: List[Tuple[str, str]] = []
+
+        # When pod in obnormal state, sometimes the conditions are not available
+        if pod_conditions:
+            for condition in pod_conditions:
+                type = condition["type"]
```
Comment (digimaun) [inline azext_edge/edge/providers/check/base/pod.py:110]: is "type" guaranteed to be in `condition`? The code on line 96 indicates conditions can be an empty array.

Comment (Elsie4ever) [reply azext_edge/edge/providers/check/base/pod.py:110]: the conditions can be empty, but when there is condition in pod_conditions(did a check in 108), the type will be there

**Review by digimaun — APPROVED** [review]

File: azext_edge/tests/edge/checks/int/helpers.py [hunk 7]
```diff
@@ -227,3 +254,12 @@ def run_check_command(
...
     return {cond["name"]: cond for cond in result["postDeployment"]}, service_present
+
+
+def condition_statuses(
+    conditions: list,
+    known_conditions: list,
+):
+    for condition in conditions:
+        if condition['type'] in known_conditions:
+            yield condition['status'] == 'True'
```
Comment (digimaun) [inline azext_edge/tests/edge/checks/int/helpers.py:265]: What is the reason for use of `yield` here ?

File: azext_edge/tests/edge/checks/int/helpers.py [hunk 8]
```diff
@@ -109,19 +108,47 @@ def assert_eval_core_service_runtime(
...
+            assert phase_conditions_eval["status"] == expected_status
 
         assert runtime_resource[namespace]["status"] == namespace_status
     assert post_deployment["evalCoreServiceRuntime"]["status"] == overall_status
 
 
+def assert_pod_conditions(pod_conditions, phase_conditions_eval, expected_status):
+    if phase_conditions_eval:
+        for condition in pod_conditions:
+            condition_type = condition.get("type")
```
Comment (digimaun) [inline azext_edge/tests/edge/checks/int/helpers.py:143]: Weren't you saying "type" is guaranteed to be there in previous comment ?

Comment (Elsie4ever) [reply azext_edge/tests/edge/checks/int/helpers.py:143]: yeah, will update to keep consistent

File: azext_edge/tests/edge/checks/int/helpers.py (hunk 7 above)
Comment (Elsie4ever) [reply azext_edge/tests/edge/checks/int/helpers.py]: updated to use return

**Pull request merged**